include/zephyr/drivers/retained_mem.h

```diff
@@ -0,0 +1,180 @@
+
+/**
+ * @typedef	retained_mem_clear_api
+ * @brief	Callback API to clear retained memory area (reset all data to 0x00).
+ * See retained_mem_clear() for argument description.
+ */
+typedef int32_t (*retained_mem_clear_api)(const struct device *dev);
+
+/**
+ * @brief Retained memory memory driver API
+ * API which can be used by a device to store data in a retained memory area. Retained memory is
+ * memory that is retained for the full power cycle of a device but is lost when power to the
+ * device is lost. This may be in a non-initialised RAM region, or in specific registers, but is
+ * not reset when a different application begins execution or the device is rebooted (without
+ * power loss). It must support byte-level reading and writing without a need to erase data before
+ * writing.
+ *
+ * Note that drivers must implement all functions, none of the functions are optional.
+ */
+struct retained_mem_driver_api {
+	retained_mem_size_api size;
+	retained_mem_read_api read;
+	retained_mem_write_api write;
+	retained_mem_clear_api clear;
+};
+
+/**
+ * @brief		Returns the size of the retained memory area.
+ *
+ * @param dev		Retained memory device to use.
+ *
+ * @retval		Positive value indicating size in bytes on success, else negative errno
+ *			code.
+ */
+__syscall int32_t retained_mem_size(const struct device *dev);
+
+static inline int32_t z_impl_retained_mem_size(const struct device *dev)
+{
+	struct retained_mem_driver_api *api = (struct retained_mem_driver_api *)dev->api;
+
+	return api->size(dev);
+}
+
+/**
+ * @brief		Reads data from the Retained memory area.
+ *
+ * @param dev		Retained memory device to use.
+ * @param offset	Offset to read data from.
+ * @param buffer	Buffer to store read data in.
+ * @param size		Size of data to read.
+ *
+ * @retval		0 on success else negative errno code.
+ */
+__syscall int32_t retained_mem_read(const struct device *dev, size_t offset, uint8_t *buffer,
+				    size_t size);
+
+static inline int32_t z_impl_retained_mem_read(const struct device *dev, size_t offset,
+					       uint8_t *buffer, size_t size)
+{
+	struct retained_mem_driver_api *api = (struct retained_mem_driver_api *)dev->api;
+
+	/* Validate user-supplied parameters */
+	if (size == 0) {
+		return -EINVAL;
+	} else if ((offset + size) > api->size(dev)) {
+		return -EINVAL;
+	}
+
+	return api->read(dev, offset, buffer, size);
+}
+
+/**
+ * @brief		Writes data to the Retained memory area - underlying data does not need to
+ *			be cleared prior to writing.
+ *
+ * @param dev		Retained memory device to use.
+ * @param offset	Offset to write data to.
+ * @param buffer	Data to write.
+ * @param size		Size of data to be written.
+ *
+ * @retval		0 on success else negative errno code.
+ */
+__syscall int32_t retained_mem_write(const struct device *dev, size_t offset,
+				     const uint8_t *buffer, size_t size);
+
+static inline int32_t z_impl_retained_mem_write(const struct device *dev, size_t offset,
```

Comment:
Shouldn't this be "plain" int?

Nonblocking: are you sure that you will not need to support devices which report partial writes?

Reply:
This is a result code, it does not return the number of bytes written, picked int32_t as that is the result code size. For this API, do not want to support partial writes, can be see as a differentiator from e.g. flash API, devices should have byte level granularity for reads/writes.

Reply:
> This is a result code, it does not return the number of bytes written

I know -- if it was a result code or number of bytes I would have said `ssize_t`.

> picked int32_t as that is the result code size.

I'm saying that `int` would seem to be a better choice for 64-bit architectures to use the natural machine word size for the return value everywhere.

Same comment applies to the read and clear APIs.

> For this API, do not want to support partial writes, can be see as a differentiator from e.g. flash API, devices should have byte level granularity for reads/writes.

ACK.

Reply:
Fixed